src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs

```diff
@@ -0,0 +1,306 @@
+                if (cursor.CurrentBlock == messageBlocks.First && cursor.IsNewestInBlock)
+                {
+                    cursor.State = Cursor.States.Idle;
+                    cursor.SequenceToken = messageBlocks.First.Value.NewestSequenceToken;
+                }
+                else // move to next
+                {
+                    int index;
+                    if (cursor.IsNewestInBlock)
+                    {
+                        cursor.CurrentBlock = cursor.CurrentBlock.Previous;
+                        cursor.CurrentBlock.Value.TryFindFirstMessage(cursor.StreamGuid, cursor.StreamNamespace, out index);
+                    }
+                    else
+                    {
+                        cursor.CurrentBlock.Value.TryFindNextMessage(cursor.Index + 1, cursor.StreamGuid, cursor.StreamNamespace, out index);
+                    }
+                    cursor.Index = index;
+                }
+
+                // check if this message is in the cursor's stream
+                if (cacheDataAdapter.IsInStream(ref currentMessage, cursor.StreamGuid, cursor.StreamNamespace))
+                {
+                    message = cacheDataAdapter.GetBatchContainer(ref currentMessage);
+                    cursor.SequenceToken = cursor.CurrentBlock.Value.GetSequenceToken(cursor.Index);
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public void Add(TQueueMessage message)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException("message");
+            }
+
+            PerformPendingPurges();
+
+            // allocate message from pool
+            CachedMessageBlock<TQueueMessage, TCachedMessage> block = pool.AllocateMessage(message);
+
+            // If new block, add message block to linked list
+            if (block != messageBlocks.FirstOrDefault())
+                messageBlocks.AddFirst(block.Node);
+
+            PerformPendingPurges();
+        }
+
+        /// <summary>
+        /// Record that we need to purge all messages associated with purge request in the next purge cycle
+        /// </summary>
+        /// <param name="purgeRequest"></param>
+        public void Purge(IDisposable purgeRequest)
+        {
+            purgeQueue.Enqueue(purgeRequest);
+        }
+
+        private void PerformPendingPurges()
+        {
+            uint purgeCount = 0;
+
+            // If there are purge requests, perform purges
+            if (!purgeQueue.IsEmpty)
+            {
+                IDisposable purgeRequest;
+                while (purgeQueue.TryDequeue(out purgeRequest))
+                {
+                    purgeCount += PerformBlockPurge(purgeRequest);
+                }
+            }
+        }
+
+        private uint PerformBlockPurge(IDisposable purgeRequest)
+        {
+            uint purgeCount = 0;
```

Comment:
It was being used for logging and backpressure, neither of which exist in the port, so I'll removed.